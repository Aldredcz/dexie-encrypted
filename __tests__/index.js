import 'fake-indexeddb/auto';
import '../test-utils/encoder';

import Dexie from 'dexie';
import 'dexie-export-import';

import nacl from 'tweetnacl';

import encrypt, { cryptoOptions } from '../index';

const keyPair = nacl.sign.keyPair.fromSeed(new Uint8Array(32));

const dbToJson = (db) => {
    return new Promise(async resolve => {
        const blob = await db.export();
        const reader = new FileReader();
        reader.addEventListener('loadend', () => {
            const data = JSON.parse(reader.result);
            resolve(data);
        });
        reader.readAsText(blob);
    });
};

describe('Encrypting', () => {
	it('should encrypt data', async (done) => {
		const db = new Dexie('MyDatabase');
		encrypt(db, keyPair.publicKey, {
			friends: cryptoOptions.DATA
		}, new Uint8Array(24))

		// Declare tables, IDs and indexes
		db.version(1).stores({
			friends: '++id, name, age'
		});
		
		await db.open();

		const original = {
			name: 'Camilla',
			age: 25,
			street: 'East 13:th Street',
			picture: 'camilla.png'
		};

		await db.friends.add(original);


		const readingDb = new Dexie('MyDatabase');
		readingDb.version(1).stores({
			friends: '++id, name, age'
		});
		await readingDb.open();

		const data = await dbToJson(readingDb);
		expect(data).toMatchSnapshot();
		done();

	});

	it('should encrypt whitelist', async (done) => {
		const db = new Dexie('whitelist');
		encrypt(db, keyPair.publicKey, {
			friends: {
				type: cryptoOptions.WHITELIST,
				fields: ['picture']
			}
		}, new Uint8Array(24))

		// Declare tables, IDs and indexes
		db.version(1).stores({
			friends: '++id, name, age'
		});
		
		await db.open();

		const original = {
			name: 'Camilla',
			age: 25,
			street: 'East 13:th Street',
			picture: 'camilla.png'
		};

		await db.friends.add(original);


		const readingDb = new Dexie('whitelist');
		readingDb.version(1).stores({
			friends: '++id, name, age'
		});
		await readingDb.open();

		const data = await dbToJson(readingDb);
		expect(data).toMatchSnapshot();
		done();

	});
	it('should encrypt blacklist', async (done) => {
		const db = new Dexie('blacklist');
		encrypt(db, keyPair.publicKey, {
			friends: {
				type: cryptoOptions.BLACKLIST,
				fields: ['street']
			}
		}, new Uint8Array(24))

		// Declare tables, IDs and indexes
		db.version(1).stores({
			friends: '++id, name, age'
		});
		
		await db.open();

		const original = {
			name: 'Camilla',
			age: 25,
			street: 'East 13:th Street',
			picture: 'camilla.png'
		};

		await db.friends.add(original);


		const readingDb = new Dexie('blacklist');
		readingDb.version(1).stores({
			friends: '++id, name, age'
		});
		await readingDb.open();

		const data = await dbToJson(readingDb);
		expect(data).toMatchSnapshot();
		done();

	});
	it('should decrypt', async (done) => {
		const db = new Dexie('decrypt-test');
		encrypt(db, keyPair.publicKey, {
			friends: cryptoOptions.DATA
		}, new Uint8Array(24))

		// Declare tables, IDs and indexes
		db.version(1).stores({
			friends: '++id, name, age'
		});
		
		await db.open();

		const original = {
			name: 'Camilla',
			age: 25,
			street: 'East 13:th Street',
			picture: 'camilla.png'
		};

		await db.friends.add({...original});

		const out = await db.friends.get(1);
		delete out.id;
		expect(original).toEqual(out);
		done();

	});

	it('should upgrade', async (done) => {
		const db = new Dexie('upgrade-db');
		encrypt(db, keyPair.publicKey, {
			friends: {
				type: cryptoOptions.WHITELIST,
				fields: ['street']
			}
		}, new Uint8Array(24))

		// Declare tables, IDs and indexes
		db.version(1).stores({
			friends: '++id, name, age'
		});
		
		await db.open();

		const original = {
			name: 'Camilla',
			age: 25,
			street: 'East 13:th Street',
			picture: 'camilla.png'
		};

		await db.friends.add(original);
		await db.close();


		const upgraded = new Dexie('upgrade-db');
		encrypt(upgraded, keyPair.publicKey, {
			friends: {
				type: cryptoOptions.WHITELIST,
				fields: ['picture']
			}
		}, new Uint8Array(24))

		upgraded.version(1).stores({
			friends: '++id, name, age'
		});

		await upgraded.open();
		
		const readingDb = new Dexie('upgrade-db');
		readingDb.version(1).stores({
			friends: '++id, name, age'
		});
		await readingDb.open();

		const data = await dbToJson(readingDb);
		expect(data).toMatchSnapshot();
		done();
	});
})