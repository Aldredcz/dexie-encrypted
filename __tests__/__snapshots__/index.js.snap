// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Encrypting should encrypt blacklist 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "$types": Object {
              "__encryptedData": "uint8array2",
              "__encryptedData.buffer": "arraybuffer",
            },
            "__encryptedData": Object {
              "buffer": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQEQDtKRiCJPL40t4EOMKkANi9XlJHDvEKVrrJnuqaZtKH+628Eh0gbW02Fl3tg==",
              "byteOffset": 0,
              "length": 70,
            },
            "age": 25,
            "id": 1,
            "name": "Camilla",
            "picture": "camilla.png",
            "street": "",
          },
        ],
        "tableName": "friends",
      },
    ],
    "databaseName": "blacklist",
    "databaseVersion": 1,
    "tables": Array [
      Object {
        "name": "friends",
        "rowCount": 1,
        "schema": "++id,name,age",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;

exports[`Encrypting should encrypt data 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "$types": Object {
              "__encryptedData": "uint8array2",
              "__encryptedData.buffer": "arraybuffer",
            },
            "__encryptedData": Object {
              "buffer": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUfsSlGjy9yeoyoKCagomeQNi6GxWHHyKKSOoDnO1cdpZAPaj/w0Fz/XkkQ8mv8HRpOEZ5iOH4z70BDC4z5JNqzsf/D0UTbCCAa5s/a5UzXgumHHTqk7lssg6Wri+vilp",
              "byteOffset": 0,
              "length": 120,
            },
            "age": 25,
            "id": 1,
            "name": "Camilla",
            "picture": "",
            "street": "",
          },
        ],
        "tableName": "friends",
      },
    ],
    "databaseName": "MyDatabase",
    "databaseVersion": 1,
    "tables": Array [
      Object {
        "name": "friends",
        "rowCount": 1,
        "schema": "++id,name,age",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;

exports[`Encrypting should encrypt whitelist 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "$types": Object {
              "__encryptedData": "uint8array2",
              "__encryptedData.buffer": "arraybuffer",
            },
            "__encryptedData": Object {
              "buffer": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUfsSlGjy9yeoyoKCagomeQNi6GxWHHyKKSOoDnO1cdpZAPaj/w0Fz/XkkQ8mv8HRpOEZ5iOH4z70BDC4z5JNqzsf/D0UTbCCAa5s/a5UzXgumHHTqk7lssg6Wri+vilp",
              "byteOffset": 0,
              "length": 120,
            },
            "age": 25,
            "id": 1,
            "name": "Camilla",
            "picture": "camilla.png",
            "street": "",
          },
        ],
        "tableName": "friends",
      },
    ],
    "databaseName": "whitelist",
    "databaseVersion": 1,
    "tables": Array [
      Object {
        "name": "friends",
        "rowCount": 1,
        "schema": "++id,name,age",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;

exports[`Encrypting should still work when you update an existing entity 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "friends": "NON_INDEXED_FIELDS",
            "id": 1,
          },
        ],
        "tableName": "_encryptionSettings",
      },
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "$types": Object {
              "__encryptedData": "uint8array2",
              "__encryptedData.buffer": "arraybuffer",
            },
            "__encryptedData": Object {
              "buffer": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMe3m+xHIk9OcimfPF78dTwNi6GxWHHyKKTOmDn+2c95bQ6Cq/RoHga+w0w0Wqt7drfla/i3g4yrlBju5wMoH+BwZ6z0FG6iMZr926voQjDE8kmOKv0usjdApEa2k+yc216yVCghRAoSNIObKyVGyCVuX29RoQnnFS9kA7IH7KA2tXAdJJrSUHMYUHGh3askAd8+/7erJoLOU3RNyDdYc0RlRXk+Mwj7mbQORHfaeeF7eUUjneuR0",
              "byteOffset": 0,
              "length": 207,
            },
            "age": 25,
            "id": 1,
            "name": "",
            "picture": "",
            "street": "",
          },
        ],
        "tableName": "friends",
      },
    ],
    "databaseName": "in-and-out-test",
    "databaseVersion": 1,
    "tables": Array [
      Object {
        "name": "_encryptionSettings",
        "rowCount": 1,
        "schema": "++id",
      },
      Object {
        "name": "friends",
        "rowCount": 1,
        "schema": "++id,age",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;

exports[`Encrypting should upgrade 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "$types": Object {
              "__encryptedData": "uint8array2",
              "__encryptedData.buffer": "arraybuffer",
            },
            "__encryptedData": Object {
              "buffer": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAPgxdjFy2otNxGdJeSMtwYANi6GxWHHyKKSOoDnO1cdpZAPaj/w0Fz/XkkQ8mv8HRpOEZ5iOH4z70BDC4z5JNqzsf/D0UTbCCAa5s/a5UzXgumHHTqk7lssg6Wri+vik4haySXEcSGg==",
              "byteOffset": 0,
              "length": 127,
            },
            "age": 25,
            "id": 1,
            "name": "Camilla",
            "picture": "camilla.png",
            "street": "",
          },
        ],
        "tableName": "friends",
      },
    ],
    "databaseName": "upgrade-db",
    "databaseVersion": 1,
    "tables": Array [
      Object {
        "name": "friends",
        "rowCount": 1,
        "schema": "++id,name,age",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;

exports[`Encrypting should wait for a promise to resolve with a key if given a promise 1`] = `
Object {
  "data": Object {
    "data": Array [
      Object {
        "inbound": true,
        "rows": Array [
          Object {
            "$types": Object {
              "__encryptedData": "uint8array2",
              "__encryptedData.buffer": "arraybuffer",
            },
            "__encryptedData": Object {
              "buffer": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUfsSlGjy9yeoyoKCagomeQNi6GxWHHyKKSOoDnO1cdpZAPaj/w0Fz/XkkQ8mv8HRpOEZ5iOH4z70BDC4z5JNqzsf/D0UTbCCAa5s/a5UzXgumHHTqk7lssg6Wri+vilp",
              "byteOffset": 0,
              "length": 120,
            },
            "age": 25,
            "id": 1,
            "name": "Camilla",
            "picture": "",
            "street": "",
          },
        ],
        "tableName": "friends",
      },
    ],
    "databaseName": "async-key",
    "databaseVersion": 1,
    "tables": Array [
      Object {
        "name": "friends",
        "rowCount": 1,
        "schema": "++id,name,age",
      },
    ],
  },
  "formatName": "dexie",
  "formatVersion": 1,
}
`;
